File: go/maximum_product_subarray.go
package leetcode

// iterative dp solution
// TC: O(n)
// SC: O(1)
func maxProduct(nums []int) int {
	n := len(nums)
	var res data
	if nums[n-1] < 0 {
		res = data{nums[n-1], 0}
	} else {
		res = data{0, nums[n-1]}
	}

	best := nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if nums[i] == 0 {
			res.negative, res.positive = 0, 0
		} else if nums[i] < 0 {
			res.negative, res.positive = min(nums[i], res.positive*nums[i]), res.negative*nums[i]
		} else {
			res.negative, res.positive = res.negative*nums[i], max(nums[i], res.positive*nums[i])
		}
		best = max(best, res.positive)
	}
	return best
}

// iterative dp solution
// TC: O(n)
// SC: O(n)
func maxProduct2(nums []int) int {
	n := len(nums)
	dp := make([]data, n)
	if nums[n-1] < 0 {
		dp[n-1] = data{nums[n-1], 0}
	} else {
		dp[n-1] = data{0, nums[n-1]}
	}

	best := nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if nums[i] == 0 {
			dp[i] = data{0, 0}
		} else if nums[i] < 0 {
			dp[i] = data{min(dp[i+1].positive*nums[i], nums[i]), nums[i] * dp[i+1].negative}
		} else {
			dp[i] = data{dp[i+1].negative * nums[i], max(nums[i]*dp[i+1].positive, nums[i])}
		}
		best = max(best, dp[i].positive)
	}
	return best
}

type data struct {
	negative int
	positive int
}

func max(x, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

func min(x, y int) int {
	if x < y {
		return x
	} else {
		return y
	}
}

File: python/maximum-product-subarray.py
from typing import List

class Solution:
    # iterative dp solution
    # TC: O(n)
    # SC: O(1)
    def maxProduct(self, nums: List[int]) -> int:
        if nums[-1] < 0:
            negative, positive = nums[-1], 0
        else:
            negative, positive = 0, nums[-1]
            
        best = nums[-1]
        for i in range(len(nums)-2, -1, -1):
            if nums[i] == 0:
                negative = positive = 0
            elif nums[i] < 0:
                negative, positive = min(nums[i], nums[i]*positive), max(nums[i], nums[i]*negative)
            else:
                negative, positive = min(nums[i], nums[i]*negative), max(nums[i], nums[i]*positive)
            best = max(best, positive)
        return best
    # iterative dp solution
    # TC: O(n)
    # SC: O(n)
    def maxProduct2(self, nums: List[int]) -> int:
        dp = [0]*len(nums)
        if nums[-1] < 0:
            dp[-1] = (0, nums[-1])
        else:
            dp[-1] = (nums[-1], 0)
        
        best = nums[-1]
        for i in range(len(nums)-2, -1, -1):
            if nums[i] == 0:
                dp[i] = (0, 0)
            elif nums[i] < 0:
                dp[i] = (max(nums[i]*dp[i+1][1], nums[i]), min(nums[i]*dp[i+1][0], nums[i]))
            else:
                dp[i] = (max(nums[i]*dp[i+1][0], nums[i]), min(nums[i]*dp[i+1][1], nums[i]))
            best = max(best, *dp[i])
        return best